File: src/test/scala/org/tindalos/principle/StructureTestManual.scala
package org.tindalos.principle

import org.tindalos.principle.domain.analyzers.structure.NodeGroup
import org.tindalos.principle.domain.analyzers.structure.{CohesiveGroupsDiscoveryModule, Graph, PackageCohesionModule, PackageStructureHints1Finder}
import org.tindalos.principle.infrastructure.service.jdepend.classdependencies.MyJDependRunner

import scala.collection.JavaConverters._

/**
 * Created by mate.magyari on 17/12/2014.
 */
object StructureTestManual extends App {

  val (targetDir, rootPackage) =
    ("//Users/mate.magyari/IdeaProjects/gamesys/gamesplatform/poker-critical-core/target/classes/"
      , "gamesys.poker.engine.model")

  /*
  val (targetDir, rootPackage) =
    ("//Users/mate.magyari/IdeaProjects/gamesys/gamesplatform/poker-player-reputation-system/target/classes/"
      , "gamesys.poker.reputation")



      val (targetDir, rootPackage) =
        ("//Users/mate.magyari/IdeaProjects/gamesys/gamesplatform/poker-game-system/target/classes/"
          , "gamesys.poker.engine")

  */


  val classes = MyJDependRunner.createNodesOfClasses(rootPackage, targetDir)
  val packages = PackageCohesionModule
    .componentsFromPackages(rootPackage, classes.asJava)
    .asScala
    .map(entry => (entry.getKey, entry.getValue))
    .toSet

  val sortedPackages = packages
    .filter(_._2.nodes.size > 1)
    .toList
    .sortBy(_._2.cohesion())


  val grouping = PackageStructureHints1Finder.makeGroups(classes)

  def aSort(s1: String, s2: String) = s1.substring(1).toInt.compareTo(s2.substring(1).toInt)

  println("\nSources:")
  grouping.labelledSources().asScala.sortBy(_.label()).foreach {
    x => println(x.label() + " -> " + x.nodeId())
  }
  println("\nGroups:")
  grouping.grouping().entrySet().asScala.foreach {
    entry => {
      println(entry.getKey.asScala.foldLeft("Sources: ")(_ + "," + _))
      entry.getValue.asScala.sorted.foreach {
        x => println("\t" + x)
      }
    }
  }

  var start = System.currentTimeMillis()

  val cohesionsForGrouping = for {
    entry <- grouping.grouping().entrySet().asScala
    nodeSet = entry.getValue.asScala.map(x => classes.find(n => n.id == x).get).toSet
    if nodeSet.size > 1
    nodeGroup = new NodeGroup(nodeSet.asJava)
  } yield (entry.getKey, nodeGroup.cohesion())

  cohesionsForGrouping.toList.sortBy(_._2).reverse.foreach {
    c => println(c._2 + " " + c._1)
  }

  //val parts = Graph.findDetachableSubgraphs(classes)
  val parts = Graph.findDetachableSubgraphs(MyJDependRunner.createNodesOfClasses("org.tindalos.principle.infrastructure").asJava)

  parts.peninsulas.asScala.foreach {
    p => {
      println("Top: " + p.frontNodes + " " + new NodeGroup(p.subgraph).cohesion())
      p.subgraph.asScala.map(_.id).toList.sorted.foreach {
        n => println("\t"+n)
      }
    }
  }

  val initialComponents = classes.map(n => new NodeGroup(java.util.Collections.singleton(n)))

  start = System.currentTimeMillis()
  val components = CohesiveGroupsDiscoveryModule.collapseToLimit(initialComponents).toList.sortBy(_.nodes.size).reverse
  println("Time1: " + (System.currentTimeMillis() - start))

  val componentsSortedBySize = components.toList.sortBy(_.nodes.size).reverse
  val componentsSortedByCohesion = components.toList.sortBy(_.cohesion()).reverse

  println("end")

}

File: src/test/java/org/tindalos/principle/StructureTestManual.java
package org.tindalos.principle;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.Map;
import java.util.Set;
import java.util.stream.Collectors;

import org.tindalos.principle.domain.analyzers.structure.CohesiveGroupsDiscoveryModule;
import org.tindalos.principle.domain.analyzers.structure.Graph;
import org.tindalos.principle.domain.core.Node;
import org.tindalos.principle.domain.analyzers.structure.NodeGroup;
import org.tindalos.principle.domain.analyzers.structure.PackageCohesionModule;
import org.tindalos.principle.domain.analyzers.structure.PackageStructureHints1Finder;
import org.tindalos.principle.infrastructure.service.jdepend.classdependencies.MyJDependRunner;

public final class StructureTestManual {

    private StructureTestManual() {
    }

    public static void main(String[] args) {
        String targetDir = "//Users/mate.magyari/IdeaProjects/gamesys/gamesplatform/poker-critical-core/target/classes/";
        String rootPackage = "gamesys.poker.engine.model";

        Set<Node> classes = MyJDependRunner.createNodesOfClasses(rootPackage, targetDir);

        Set<Map.Entry<String, NodeGroup>> packages = PackageCohesionModule.componentsFromPackages(rootPackage, classes);

        var sortedPackages = packages.stream()
                .filter(entry -> entry.getValue().nodes().size() > 1)
                .sorted(Comparator.comparingDouble(entry -> entry.getValue().cohesion()))
                .toList();

        var grouping = PackageStructureHints1Finder.makeGroups(classes);

        System.out.println("\nSources:");
        grouping.labelledSources().stream()
                .sorted(Comparator.comparing(source -> source.label()))
                .forEach(source -> System.out.println(source.label() + " -> " + source.nodeId()));

        System.out.println("\nGroups:");
        grouping.grouping().forEach((key, value) -> {
            System.out.println(key.stream().reduce("Sources: ", (acc, item) -> acc + "," + item));
            value.stream().sorted().forEach(nodeId -> System.out.println("\t" + nodeId));
        });

        long start = System.currentTimeMillis();

        Map<String, Node> classesById = classes.stream().collect(Collectors.toMap(Node::id, node -> node));

        var cohesionsForGrouping = grouping.grouping().entrySet().stream()
                .map(entry -> {
                    Set<Node> nodeSet = entry.getValue().stream().map(classesById::get).collect(Collectors.toSet());
                    return Map.entry(entry.getKey(), new NodeGroup(nodeSet).cohesion());
                })
                .filter(entry -> entry.getKey().size() > 1)
                .toList();

        cohesionsForGrouping.stream()
                .sorted(Comparator.comparingDouble((Map.Entry<Set<String>, Double> entry) -> entry.getValue()).reversed())
                .forEach(entry -> System.out.println(entry.getValue() + " " + entry.getKey()));

        var parts = Graph.findDetachableSubgraphs(MyJDependRunner.createNodesOfClasses("org.tindalos.principle.infrastructure"));
        parts.peninsulas().forEach(peninsula -> {
            System.out.println("Top: " + peninsula.frontNodes() + " " + new NodeGroup(peninsula.subgraph()).cohesion());
            peninsula.subgraph().stream().map(Node::id).sorted().forEach(nodeId -> System.out.println("\t" + nodeId));
        });

        Set<NodeGroup> initialComponents = classes.stream()
                .map(node -> new NodeGroup(Collections.singleton(node)))
                .collect(Collectors.toSet());

        start = System.currentTimeMillis();
        var components = new ArrayList<>(CohesiveGroupsDiscoveryModule.collapseToLimit(initialComponents));
        components.sort(Comparator.comparingInt((NodeGroup group) -> group.nodes().size()).reversed());
        System.out.println("Time1: " + (System.currentTimeMillis() - start));

        var componentsSortedBySize = components.stream()
                .sorted(Comparator.comparingInt((NodeGroup group) -> group.nodes().size()).reversed())
                .toList();
        var componentsSortedByCohesion = components.stream()
                .sorted(Comparator.comparingDouble((NodeGroup group) -> group.cohesion()).reversed())
                .toList();

        System.out.println(sortedPackages.size() + componentsSortedBySize.size() + componentsSortedByCohesion.size());
        System.out.println("end");
    }
}
